sdk/servicebus/Azure.Messaging.ServiceBus/tests/ReceiverLiveTests.cs

```diff
@@ -111,5 +113,112 @@ await foreach (var item in receiver.ReceiveBatchAsync(messageCount))
                 Assert.IsNull(message.Result);
             }
         }
+
+        [Test]
+        [TestCase(1)]
+        [TestCase(5)]
+        [TestCase(10)]
+        [TestCase(20)]
+        public async Task Receive_Event(int numThreads)
+        {
+            await using (var scope = await ServiceBusScope.CreateWithQueue(
+                enablePartitioning: false,
+                enableSession: false))
+            {
+                await using var sender = new ServiceBusSenderClient(
```

Comment:
This approach feels brittle;  there's a lot of non-determinism in timing for the messaging services and within our CI environment.   Most of the flaky tests we've had in Event Hubs are due to this kind of delay/timing.

We've started to try and use task completion source instances that we can await instead, so that the coordination is deterministic.   In this case, maybe consider creating a pool of TCS instances and awaiting them all?    For each tick, you can set one and then advance an index tracker?    Something like:

```csharp
var completionSources = Enumerable
    .Range(0, numThreads)
    .Select(index => new TaskCompletionSource<bool>())
    .ToArray();

var completionSourceIndex = 0;

// .... SNIP ...
async Task ProcessMessage(ServiceBusMessage message)
{
    await receiver.CompleteAsync(message.SystemProperties.LockToken);
    Interlocked.Increment(ref messageCt);
    var setIndex = completionSourceIndex++;

    completionSources[setIndex].TrySet(true);
}

//... SNIP ...
await receiver.StartProcessingAsync(options);
await Task.WhenAll(completionSources);

Assert.That(messageCt, Is.EqualTo(numThreads).Within(2));
```

Reply:
Thanks, this is great!

Reply:
Just one thing, TaskCompletionSource should always be created with TaskCreationOptions.RunContinuationsAsynchronously to avoid deadlocks caused by continuations running inline.